datadog/dogstatsd/base.py

```diff
@@ -47,17 +47,21 @@ def __exit__(self, type, value, traceback):
         self.close_buffer()
 
     def get_socket(self):
-        '''
-        Return a connected socket
-        '''
+        """
+        Return a connected socket.
+
+        Note: connect the socket before assigning it to the class instance to
+        avoid bad thread race conditions.
+        """
         if not self.socket:
-            self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
-            self.socket.connect((self.host, self.port))
+            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            sock.connect((self.host, self.port))
+            self.socket = sock
         return self.socket
```

Comment:
I'm seeing these races in a newer version of this module. Pretty sure there is still a race in this code. If another thread closes the socket between line 59 and 60, `get_socket` could return `None` or a closed socket. Similar race in `close_socket` itself.